### ISSUE TITLE
2CR and regression failures due to underlying system libraries (BLAS?)

### ISSUE THREAD

**Issue by lintool** [castorini/pyserini#1630]
Prior to release of v0.22.0, I had verified that all 2CRs and regressions passed on `tuna`. However, when I installed the PyPI artifact on `linux.cs`, I observed some failures. Of the PyPI 2CRs:

+ `msmarco-v1-passage`
+ `msmarco-v1-doc`
+ `msmarco-v2-passage`
+ `msmarco-v2-doc`
+ `miracl`
+ `mrtydi`

The MS MARCO ones passed fine, but failures occurred with `miracl` and `mrtydi`. After some digging around, I traced it back to OS system libraries - `tuna` was on Ubuntu 18, while `linux.cs` is on Ubtuntu 22. I suspect, but haven't confirmed, that the issue can be traced back to differences in the BLAS libraries, since the failures were on the Faiss runs.

Today, `tuna` was upgraded to Ubuntu 22. Will follow up on the situation.

**Comment by lintool** [comment]
I've posted the regression failures I've observed in #1631

**Comment by lintool** [comment]
These are the test failures:

## `miracl`

```
condition mdpr-tied-pft-msmarco.bn:
  - split: train

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics tools/topics-and-qrels/topics.miracl-v1.0-bn-train.tsv \
  --index miracl-v1.0-bn-mdpr-tied-pft-msmarco \
  --output runs/run.miracl.mdpr-tied-pft-msmarco.bn.train.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.4589 [FAIL] expected 0.4362
      R@100  : 0.8244 [FAIL] expected 0.8045
  - split: dev

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics miracl-v1.0-bn-dev \
  --index miracl-v1.0-bn-mdpr-tied-pft-msmarco \
  --output runs/run.miracl.mdpr-tied-pft-msmarco.bn.dev.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.4622 [FAIL] expected 0.4427
      R@100  : 0.8311 [FAIL] expected 0.8193

condition mdpr-tied-pft-msmarco.hi:
  - split: train

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics tools/topics-and-qrels/topics.miracl-v1.0-hi-train.tsv \
  --index miracl-v1.0-hi-mdpr-tied-pft-msmarco \
  --output runs/run.miracl.mdpr-tied-pft-msmarco.hi.train.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.3700 [FAIL] expected 0.3685
      R@100  : 0.7760 [FAIL] expected 0.7780
  - split: dev

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics miracl-v1.0-hi-dev \
  --index miracl-v1.0-hi-mdpr-tied-pft-msmarco \
  --output runs/run.miracl.mdpr-tied-pft-msmarco.hi.dev.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.3895 [FAIL] expected 0.3830
      R@100  : 0.7810 [FAIL] expected 0.7755

condition mdpr-tied-pft-msmarco.te:
  - split: train

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics tools/topics-and-qrels/topics.miracl-v1.0-te-train.tsv \
  --index miracl-v1.0-te-mdpr-tied-pft-msmarco \
  --output runs/run.miracl.mdpr-tied-pft-msmarco.te.train.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.3724 [FAIL] expected 0.3723
      R@100  : 0.7698 [OK]

condition mdpr-tied-pft-msmarco-ft-all.bn:
  - split: train

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics tools/topics-and-qrels/topics.miracl-v1.0-bn-train.tsv \
  --index miracl-v1.0-bn-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.miracl.mdpr-tied-pft-msmarco-ft-all.bn.train.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.6863 [FAIL] expected 0.6823
      R@100  : 0.8706 [FAIL] expected 0.8646
  - split: dev

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics miracl-v1.0-bn-dev \
  --index miracl-v1.0-bn-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.miracl.mdpr-tied-pft-msmarco-ft-all.bn.dev.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.5923 [FAIL] expected 0.5804
      R@100  : 0.8660 [FAIL] expected 0.8480

condition mdpr-tied-pft-msmarco-ft-all.hi:
  - split: train

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics tools/topics-and-qrels/topics.miracl-v1.0-hi-train.tsv \
  --index miracl-v1.0-hi-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.miracl.mdpr-tied-pft-msmarco-ft-all.hi.train.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.3363 [FAIL] expected 0.3336
      R@100  : 0.6408 [FAIL] expected 0.6388
  - split: dev

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics miracl-v1.0-hi-dev \
  --index miracl-v1.0-hi-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.miracl.mdpr-tied-pft-msmarco-ft-all.hi.dev.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.3324 [FAIL] expected 0.3286
      R@100  : 0.6348 [FAIL] expected 0.6371

condition mdpr-tied-pft-msmarco-ft-all.te:
  - split: train

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics tools/topics-and-qrels/topics.miracl-v1.0-te-train.tsv \
  --index miracl-v1.0-te-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.miracl.mdpr-tied-pft-msmarco-ft-all.te.train.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.8756 [FAIL] expected 0.8757
      R@100  : 0.9725 [OK]

condition mdpr-tied-pft-msmarco-ft-all.th:
  - split: train

python -m pyserini.search.faiss \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics tools/topics-and-qrels/topics.miracl-v1.0-th-train.tsv \
  --index miracl-v1.0-th-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.miracl.mdpr-tied-pft-msmarco-ft-all.th.train.top1000.txt \
  --batch 128 --threads 16 --hits 1000

      nDCG@10: 0.7761 [OK]
      R@100  : 0.9240 [FAIL] expected 0.9241

condition bm25-mdpr-tied-pft-msmarco-hybrid.bn:
  - split: train

python -m pyserini.fusion \
  --runs  runs/run.miracl.bm25.bn.train.top1000.txt runs/run.miracl.mdpr-tied-pft-msmarco.bn.train.top1000.txt \
  --output runs/run.miracl.bm25-mdpr-tied-pft-msmarco-hybrid.bn.train.top1000.txt --method interpolation --alpha 0.5 --depth 1000 --k 1000

      nDCG@10: 0.6733 [FAIL] expected 0.6587
      R@100  : 0.9394 [FAIL] expected 0.9297
  - split: dev

python -m pyserini.fusion \
  --runs  runs/run.miracl.bm25.bn.dev.top1000.txt runs/run.miracl.mdpr-tied-pft-msmarco.bn.dev.top1000.txt \
  --output runs/run.miracl.bm25-mdpr-tied-pft-msmarco-hybrid.bn.dev.top1000.txt --method interpolation --alpha 0.5 --depth 1000 --k 1000

      nDCG@10: 0.6714 [FAIL] expected 0.6540
      R@100  : 0.9491 [FAIL] expected 0.9321

condition bm25-mdpr-tied-pft-msmarco-hybrid.hi:
  - split: train

python -m pyserini.fusion \
  --runs  runs/run.miracl.bm25.hi.train.top1000.txt runs/run.miracl.mdpr-tied-pft-msmarco.hi.train.top1000.txt \
  --output runs/run.miracl.bm25-mdpr-tied-pft-msmarco-hybrid.hi.train.top1000.txt --method interpolation --alpha 0.5 --depth 1000 --k 1000

      nDCG@10: 0.6187 [FAIL] expected 0.6217
      R@100  : 0.9038 [FAIL] expected 0.9059
  - split: dev

python -m pyserini.fusion \
  --runs  runs/run.miracl.bm25.hi.dev.top1000.txt runs/run.miracl.mdpr-tied-pft-msmarco.hi.dev.top1000.txt \
  --output runs/run.miracl.bm25-mdpr-tied-pft-msmarco-hybrid.hi.dev.top1000.txt --method interpolation --alpha 0.5 --depth 1000 --k 1000

      nDCG@10: 0.6145 [FAIL] expected 0.6157
      R@100  : 0.9153 [FAIL] expected 0.9115
```

## `mrtydi`

```
condition mdpr-tied-pft-msmarco-ft-all.bn:
  - split: train

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics mrtydi-v1.1-bengali-train \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.mrtydi.mdpr-tied-pft-msmarco-ft-all.bn.train.txt --hits 100

      MRR@100: 0.9540 [FAIL] expected 0.9620
      R@100  : 0.9994 [FAIL] expected 1.0000
  - split: dev

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics mrtydi-v1.1-bengali-dev \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.mrtydi.mdpr-tied-pft-msmarco-ft-all.bn.dev.txt --hits 100

      MRR@100: 0.6086 [FAIL] expected 0.5897
      R@100  : 0.9125 [FAIL] expected 0.8977
  - split: test

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics mrtydi-v1.1-bengali-test \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.mrtydi.mdpr-tied-pft-msmarco-ft-all.bn.test.txt --hits 100

      MRR@100: 0.6427 [FAIL] expected 0.6228
      R@100  : 0.9550 [OK]

condition mdpr-tied-pft-msmarco-ft-all.te:
  - split: train

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco-ft-all \
  --topics mrtydi-v1.1-telugu-train \
  --index mrtydi-v1.1-telugu-mdpr-tied-pft-msmarco-ft-all \
  --output runs/run.mrtydi.mdpr-tied-pft-msmarco-ft-all.te.train.txt --hits 100

      MRR@100: 0.9676 [FAIL] expected 0.9679
      R@100  : 1.0000 [OK]

condition mdpr-tied-pft-msmarco.bn:
  - split: train

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics mrtydi-v1.1-bengali-train \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-msmarco \
  --output runs/run.mrtydi.mdpr-tied-pft-msmarco.bn.train.txt --hits 100

      MRR@100: 0.3051 [FAIL] expected 0.2920
      R@100  : 0.7481 [FAIL] expected 0.7323
  - split: dev

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics mrtydi-v1.1-bengali-dev \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-msmarco \
  --output runs/run.mrtydi.mdpr-tied-pft-msmarco.bn.dev.txt --hits 100

      MRR@100: 0.3150 [FAIL] expected 0.2993
      R@100  : 0.7341 [FAIL] expected 0.7318
  - split: test

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics mrtydi-v1.1-bengali-test \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-msmarco \
  --output runs/run.mrtydi.mdpr-tied-pft-msmarco.bn.test.txt --hits 100

      MRR@100: 0.4171 [FAIL] expected 0.3969
      R@100  : 0.8198 [FAIL] expected 0.7838

condition mdpr-tied-pft-msmarco.te:
  - split: train

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-msmarco \
  --topics mrtydi-v1.1-telugu-train \
  --index mrtydi-v1.1-telugu-mdpr-tied-pft-msmarco \
  --output runs/run.mrtydi.mdpr-tied-pft-msmarco.te.train.txt --hits 100

      MRR@100: 0.3060 [FAIL] expected 0.3059
      R@100  : 0.7513 [FAIL] expected 0.7510

condition mdpr-tied-pft-nq.bn:
  - split: train

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-nq \
  --topics mrtydi-v1.1-bengali-train \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-nq \
  --output runs/run.mrtydi.mdpr-tied-pft-nq.bn.train.txt --hits 100

      MRR@100: 0.2475 [FAIL] expected 0.2371
      R@100  : 0.6538 [FAIL] expected 0.6281
  - split: dev

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-nq \
  --topics mrtydi-v1.1-bengali-dev \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-nq \
  --output runs/run.mrtydi.mdpr-tied-pft-nq.bn.dev.txt --hits 100

      MRR@100: 0.2528 [FAIL] expected 0.2414
      R@100  : 0.6636 [FAIL] expected 0.6409
  - split: test

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-nq \
  --topics mrtydi-v1.1-bengali-test \
  --index mrtydi-v1.1-bengali-mdpr-tied-pft-nq \
  --output runs/run.mrtydi.mdpr-tied-pft-nq.bn.test.txt --hits 100

      MRR@100: 0.2546 [FAIL] expected 0.2535
      R@100  : 0.7162 [FAIL] expected 0.7072

condition mdpr-tied-pft-nq.te:
  - split: train

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder-class auto \
  --encoder castorini/mdpr-tied-pft-nq \
  --topics mrtydi-v1.1-telugu-train \
  --index mrtydi-v1.1-telugu-mdpr-tied-pft-nq \
  --output runs/run.mrtydi.mdpr-tied-pft-nq.te.train.txt --hits 100

      MRR@100: 0.1484 [FAIL] expected 0.1483
      R@100  : 0.4162 [OK]

condition mdpr-split-pft-nq.bn:
  - split: train

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder castorini/mdpr-question-nq \
  --topics mrtydi-v1.1-bengali-train \
  --index mrtydi-v1.1-bengali-mdpr-nq \
  --output runs/run.mrtydi.mdpr-split-pft-nq.bn.train.txt --hits 100

      MRR@100: 0.2402 [FAIL] expected 0.2293
      R@100  : 0.6614 [FAIL] expected 0.6454
  - split: dev

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder castorini/mdpr-question-nq \
  --topics mrtydi-v1.1-bengali-dev \
  --index mrtydi-v1.1-bengali-mdpr-nq \
  --output runs/run.mrtydi.mdpr-split-pft-nq.bn.dev.txt --hits 100

      MRR@100: 0.2578 [FAIL] expected 0.2367
      R@100  : 0.6750 [FAIL] expected 0.6511
  - split: test

python -m pyserini.search.faiss --threads 16 --batch-size 512 \
  --encoder castorini/mdpr-question-nq \
  --topics mrtydi-v1.1-bengali-test \
  --index mrtydi-v1.1-bengali-mdpr-nq \
  --output runs/run.mrtydi.mdpr-split-pft-nq.bn.test.txt --hits 100

      MRR@100: 0.2962 [FAIL] expected 0.2911
      R@100  : 0.7928 [FAIL] expected 0.7793
```

**Comment by lintool** [comment]
Closed by #1632
